src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1609,6 +1692,139 @@ RebalanceTableShards(RebalanceOptions *options, Oid shardReplicationModeOid)
 }
 
 
+static bool
+HasConcurrentRebalance(RebalanceOptions *options)
+{
+	if (HasNonTerminalJobOfType("rebalance", NULL))
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+static void
+ErrorOnConcurrentRebalance(RebalanceOptions *options)
+{
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, options->relationIdList)
+	{
+		/* this provides the legacy error when the lock can't be acquired */
+		AcquireRebalanceColocationLock(relationId, options->operationName);
+	}
+
+	if (HasConcurrentRebalance(options))
+	{
+		/* TODO find hint/detail messages to show */
+		ereport(ERROR, (errmsg("a rebalance is already running")));
+	}
+}
+
+
+/*
+ * RebalanceTableShardsBackground rebalances the shards for the relations
+ * inside the relationIdList across the different workers. It does so using our
+ * background job+task infrastructure.
+ */
+static void
+RebalanceTableShardsBackground(RebalanceOptions *options, Oid shardReplicationModeOid)
+{
+	if (list_length(options->relationIdList) == 0)
+	{
+		return;
+	}
+
+	char *operationName = "rebalance";
+	if (options->drainOnly)
+	{
+		operationName = "move";
+	}
+
+	options->operationName = operationName;
+	ErrorOnConcurrentRebalance(options);
+
+	List *placementUpdateList = GetRebalanceSteps(options);
+
+	if (list_length(placementUpdateList) == 0)
+	{
+		return;
+	}
+
+	/* find the name of the shard transfer mode to interpolate in the scheduled command */
+	Datum shardTranferModeLabelDatum =
+		DirectFunctionCall1(enum_out, shardReplicationModeOid);
+	char *shardTranferModeLabel = DatumGetCString(shardTranferModeLabelDatum);
+
+	if (list_length(placementUpdateList) == 0)
+	{
+		return;
+	}
+
+	/* schedule planned moves */
+	int64 jobId = CreateBackgroundJob("rebalance", "Rebalance all colocation groups");
+
+	/* find all copy commands to be executed before we can move any shard to new worker */
+	const char shardTransferMode = LookupShardTransferMode(shardReplicationModeOid);
+	List *referenceTableCopyCommands =
+		GetShardCopyCommandsForMissingReferenceTablePlacements(shardTransferMode);
+
+	/* keep track of the copy commands as dependencies for subsequent moves */
+	int maxDependencies = 1 + list_length(referenceTableCopyCommands);
+	int64 *prevJobId = palloc0(sizeof(int64) * maxDependencies);
+	int prevJobIdx = 0;
+
+	/* schedule tasks for all copies before we move any shards */
+	char *referenceTableCopyCommand = NULL;
+	foreach_ptr(referenceTableCopyCommand, referenceTableCopyCommands)
+	{
+		/*
+		 * Create copy tasks without adding them as dependencies to each other, they will
```

Comment:
that sounds nice: so a failure in one of them will allow the others to continue?

Reply:
yes, the copies don't get cancelled after a failure, but because any move depends on all copies they won't start and the job will transition to failed after all copies either completed or failed.